File: src/frontend/src/api/aiReview.ts
import { api } from "./request"
import type {
  AIReviewResult,
  AIReviewRequest,
  AIChapterInspectionResponse,
  CheckerBaseResponse,
  ContinuityCheckResponse,
  PacingCheckResponse,
  OOCCheckResponse,
  HighPointCheckResponse,
  ReaderPullCheckResponse,
  ExtractedEntity,
} from "./types"

// ============================================
// AI Review API
// ============================================

export const aiReviewApi = {
  /**
   * Review world settings for consistency using AI.
   * Analyzes characters, locations, items, factions, and rules
   * for logical consistency and potential issues.
   */
  reviewSettings: async (
    data: AIReviewRequest
  ): Promise<AIReviewResult> => {
    return api.post<AIReviewResult>("/ai/review", data)
  },

  /**
   * Run AI inspection on a chapter.
   * Checks for consistency, plot holes, character consistency, etc.
   */
  inspectChapter: async (
    chapterId: number
  ): Promise<AIChapterInspectionResponse> => {
    return api.post<AIChapterInspectionResponse>(`/ai/chapters/${chapterId}/inspect`)
  },

  /**
   * Extract entities from chat messages.
   * Returns extracted characters, locations, items, factions, etc.
   */
  extractEntities: async (
    chatMessages: Array<{ role: string; content: string }>
  ): Promise<{ entities: ExtractedEntity[] }> => {
    return api.post<{ entities: ExtractedEntity[] }>("/ai/extract-entities", {
      chat_messages: chatMessages,
    })
  },
}

// ============================================
// AI Checker API
// ============================================

export const checkerApi = {
  /**
   * Check world consistency for a chapter.
   * Validates locations, timelines, power levels, item ownership,
   * and faction relationships against established world settings.
   */
  checkConsistency: async (
    chapterId: number
  ): Promise<CheckerBaseResponse> => {
    return api.post<CheckerBaseResponse>("/ai/check/consistency", {
      chapter_id: chapterId,
    })
  },

  /**
   * Check scene and narrative continuity for a chapter.
   * Validates scene transitions, event consistency, character state continuity,
   * plot thread fulfillment, and detail coherence with previous chapters.
   */
  checkContinuity: async (
    chapterId: number
  ): Promise<ContinuityCheckResponse> => {
    return api.post<ContinuityCheckResponse>("/ai/check/continuity", {
      chapter_id: chapterId,
    })
  },

  /**
   * Check narrative pacing and strand ratios for a chapter.
   * Analyzes quest/fire/constellation strand ratios against the
   * target 60%/20%/20% distribution.
   */
  checkPacing: async (
    chapterId: number
  ): Promise<PacingCheckResponse> => {
    return api.post<PacingCheckResponse>("/ai/check/pacing", {
      chapter_id: chapterId,
    })
  },

  /**
   * Check for Out-Of-Character behavior.
   * Validates that a character's actions in the chapter are consistent
   * with their established personality, desires, and flaws.
   */
  checkOOC: async (
    chapterId: number,
    characterId: number
  ): Promise<OOCCheckResponse> => {
    return api.post<OOCCheckResponse>("/ai/check/ooc", {
      chapter_id: chapterId,
      character_id: characterId,
    })
  },

  /**
   * Check excitement density and high points for a chapter.
   * Analyzes climax distribution, emotional pacing, buildup adequacy,
   * and chapter-ending hook strength.
   */
  checkHighPoint: async (
    chapterId: number
  ): Promise<HighPointCheckResponse> => {
    return api.post<HighPointCheckResponse>("/ai/check/high-point", {
      chapter_id: chapterId,
    })
  },

  /**
   * Check reader engagement and hooks for a chapter.
   * Analyzes opening hooks, ending suspense, conflict drivers,
   * curiosity gaps, and emotional resonance points.
   */
  checkReaderPull: async (
    chapterId: number
  ): Promise<ReaderPullCheckResponse> => {
    return api.post<ReaderPullCheckResponse>("/ai/check/reader-pull", {
      chapter_id: chapterId,
    })
  },

  /**
   * Run all checkers on a chapter.
   * Convenience method that runs consistency, continuity, pacing,
   * high-point, and reader-pull checks.
   */
  checkAll: async (
    chapterId: number
  ): Promise<{
    consistency: CheckerBaseResponse
    continuity: ContinuityCheckResponse
    pacing: PacingCheckResponse
    highPoint: HighPointCheckResponse
    readerPull: ReaderPullCheckResponse
  }> => {
    const [consistency, continuity, pacing, highPoint, readerPull] =
      await Promise.all([
        checkerApi.checkConsistency(chapterId),
        checkerApi.checkContinuity(chapterId),
        checkerApi.checkPacing(chapterId),
        checkerApi.checkHighPoint(chapterId),
        checkerApi.checkReaderPull(chapterId),
      ])
    return { consistency, continuity, pacing, highPoint, readerPull }
  },
}

// Export all AI review APIs
export default {
  review: aiReviewApi,
  checker: checkerApi,
}

File: src/backend/api/v1/endpoints/ai/ai_review.py
# Auto Novel Writer - AI Review & Extract Endpoints
# POST /review, POST /extract-entities, POST /chapters/{chapter_id}/inspect

from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field, field_validator

from backend.core.services.chapter.chapter_service import ChapterService

from .dependencies import (
    get_ai_service,
    get_chapter_service,
)

router = APIRouter()


# Request/Response models

class ReviewRequest(BaseModel):
    """Request for AI setting review."""
    model_config = {"json_schema_extra": {
        "example": {
            "settings_data": {"characters": [{"name": "主角", "personality": "冷静"}]}
        }
    }}

    settings_data: dict = Field(..., description="设定数据字典")

    @field_validator('settings_data')
    @classmethod
    def validate_settings_data(cls, v: dict) -> dict:
        if not v:
            raise ValueError('settings_data cannot be empty')
        return v


class ReviewResponse(BaseModel):
    """Response for AI review."""
    model_config = {"json_schema_extra": {
        "example": {
            "review_content": "设定审查结果...",
            "raw_response": {}
        }
    }}

    review_content: str = Field(..., description="审查结果文本")
    raw_response: dict = Field(..., description="原始AI响应")


class ExtractEntitiesRequest(BaseModel):
    """Request to extract entities from chat messages."""
    model_config = {"json_schema_extra": {
        "example": {
            "chat_messages": [{"role": "user", "content": "主角叫张三"}]
        }
    }}

    chat_messages: list = Field(..., description="聊天消息列表")

    @field_validator('chat_messages')
    @classmethod
    def validate_chat_messages(cls, v: list) -> list:
        if not v:
            raise ValueError('chat_messages cannot be empty')
        return v


# Endpoints

@router.post(
    "/review",
    summary="AI审查设定",
    description="使用AI审查世界设定的一致性，分析角色、地点、物品、势力、规则之间的逻辑一致性。",
)
async def review_settings(request: ReviewRequest) -> ReviewResponse:
    """Review world settings for consistency using AI.

    Analyzes characters, locations, items, factions, and rules
    for logical consistency and potential issues.
    """
    ai_service = get_ai_service()

    result = await ai_service.review_settings(request.settings_data)
    return ReviewResponse(
        review_content=result["review_content"],
        raw_response=result["raw_response"]
    )


@router.post(
    "/extract-entities",
    summary="从聊天中提取实体",
    description="从聊天消息中提取角色、地点、物品、势力等实体信息。",
)
async def extract_entities(request: ExtractEntitiesRequest) -> dict:
    """Extract entities from chat messages.

    Returns extracted characters, locations, items, factions, etc.
    """
    ai_service = get_ai_service()

    entities = await ai_service.extract_entities(request.chat_messages)
    return {"entities": entities}


@router.post(
    "/chapters/{chapter_id}/inspect",
    summary="AI审查章节",
    description="对指定章节进行AI审查，检查角色一致性、情节逻辑、世界观一致性、伏笔运用等。",
)
async def inspect_chapter(
    chapter_id: int,
    chapter_service: ChapterService = Depends(get_chapter_service),
) -> dict:
    """Run AI inspection on a chapter.

    Checks for consistency, plot holes, character consistency, etc.
    """
    # Get chapter
    chapter = await chapter_service.get_chapter(chapter_id)
    if not chapter:
        raise HTTPException(status_code=404, detail="Chapter not found")

    # Get latest draft for this chapter
    drafts = await chapter_service.list_draft_versions(chapter_id)
    draft = max(drafts, key=lambda d: d.version_number, default=None) if drafts else None

    if not draft:
        raise HTTPException(status_code=404, detail="No draft found for chapter")

    ai_service = get_ai_service()

    # Build inspection prompt
    inspection_prompt = f"""请审查以下章节内容：

标题：{chapter.title or '无标题'}
大纲摘要：{chapter.summary or '无'}
章节内容：
{draft.content}

请检查以下方面：
1. 角色一致性（性格、行为是否矛盾）
2. 情节逻辑（是否有漏洞）
3. 世界观一致性（设定是否冲突）
4. 伏笔运用（伏笔是否有照应）
5. 建议优化

请以JSON格式返回，包含issues和suggestions字段。"""

    review_result = await ai_service.review_settings({"content": inspection_prompt})

    return {
        "chapter_id": chapter_id,
        "review_content": review_result["review_content"],
        "raw_response": review_result["raw_response"]
    }
